lib/src/metta/runner/stdlib.rs

```diff
@@ -1012,6 +948,302 @@ impl Grounded for MatchOp {
+        }
+    }
+
+    impl Grounded for GitModuleOp {
+        fn type_(&self) -> Atom {
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_ATOM, UNIT_TYPE()])
+        }
+
+        fn execute(&self, args: &[Atom]) -> Result<Vec<Atom>, ExecError> {
+            let arg_error = "git-module! expects a URL; use quotes if needed";
+            let url_arg_atom = args.get(0).ok_or_else(|| ExecError::from(arg_error))?;
+            // TODO: When we figure out how to address varargs, it will be nice to take an optional branch name
+
+            let url = match url_arg_atom {
+                Atom::Symbol(url_arg) => url_arg.name(),
+                Atom::Grounded(g) => g.downcast_ref::<Str>().ok_or_else(|| ExecError::from(arg_error))?.as_str(),
+                _ => return Err(arg_error.into()),
+            };
+            let url = strip_quotes(url);
+
+            // TODO: Depending on what we do with `register-module!`, we might want to let the
+            // caller provide an optional mod_name here too, rather than extracting it from the url
+            let mod_name = match mod_name_from_url(url) {
+                Some(mod_name) => mod_name,
+                None => return Err(ExecError::from("git-module! error extracting module name from URL"))
+            };
+
+            let ctx_ref = self.context.lock().unwrap().last().unwrap().clone();
+            let mut context = ctx_ref.lock().unwrap();
+
+            let git_mod_location = ModuleGitLocation::new(url.to_string());
+
+            if let Some((loader, descriptor)) = git_mod_location.get_loader_in_explicit_catalog(&mod_name, UpdateMode::TryFetchLatest, context.metta.environment()).map_err(|e| ExecError::from(e))? {
+                context.get_or_init_module_with_descriptor(&mod_name, descriptor, loader).map_err(|e| ExecError::from(e))?;
+            }
+
+            unit_result()
+        }
+
+        fn match_(&self, other: &Atom) -> MatchResultIter {
+            match_by_equality(self, other)
+        }
+    }
+
+    /// Lists contents of all Catalogs that support the "list" method
+    #[derive(Clone, Debug)]
+    pub struct CatalogListOp {
+        metta: Metta
+    }
+
+    impl PartialEq for CatalogListOp {
+        fn eq(&self, _other: &Self) -> bool { true }
+    }
+
+    impl CatalogListOp {
+        pub fn new(metta: Metta) -> Self {
+            Self{ metta }
+        }
+    }
+
+    impl Display for CatalogListOp {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "catalog-list")
+        }
+    }
+
+    impl Grounded for CatalogListOp {
+        fn type_(&self) -> Atom {
+            //TODO-FUTURE, when we decide on a friendly standard for var-args, it would be nice to
+            // allow an optional arg to list a specific catalog.  For now we list all of them
+            //TODO-FUTURE, we may want to return the list as atoms, but now it just prints to stdout
+            Atom::expr([ARROW_SYMBOL, UNIT_TYPE()])
```

Comment:
> Regarding listing modules vs listing the catalog contents, they are different. The analogy is a module being present in a managed catalog is akin to it being "installed". ie. it's available for loading, but it's not necessarily loaded. Catalogs are about module access and version-control.

I was writing about modules as a catalog contents not about loaded modules. Thus `(module-list <catalog>)` command in my comment is for listing modules within catalog not loaded modules.

> Regarding the ability to list the available catalogs, do you think it's needed? there will only ever bee a handful of hard-coded ones, like "git-modules", and whatever the user added in their environment.metta cfg file.

I was thinking about making an API which can be used by program to get a structured view on catalogs and modules inside them. From this perspective I think having a function to list available catalogs is useful.

> My thinking was that this is a good candidate to return results via atom-serde formatted atoms.

I am sorry if you already explained it somewhere and I forgot about it but what is `atom-serde` formatted atoms?

Reply:
> I was thinking about making an API which can be used by program to get a structured view on catalogs and modules inside them.

I added a discussion about the interface here: 6ffe2f9  All of the interfaces you want (and more) are easy enough to add, but we get into some conceptually difficult territory because sometimes distinctions aren't clear-cut.

> what is atom-serde formatted atoms

I've been thinking for a while that: 1. a Space is a type of database, 2. "lowest-common-denominator" schemas, such as Json, can encode most types of structured data.  So, if we can define a convention for conversion of any structure to and from atoms, we can perform bidirectional conversion on any data structure we want "for free"*.  https://github.com/trueagi-io/hyperon-experimental/issues/455

*for free in terms of implementing the conversion code, but also in terms of providing utility functions to work with the data, and in terms of documenting the format and educating users about how to access it.